fuzzing/valuegenerationtracer/valuegeneration_tracer.go

```diff
@@ -0,0 +1,357 @@
+func (t *ValueGenerationTracer) getCallFrameReturnValue() TransactionOutputValues {
+	// Define some strings that represent our current call frame
+	var method *abi.Method
+
+	// Define a slice of any to represent return values of the current call frame
+	var outputValues TransactionOutputValues
+
+	// Resolve our method definition
+	if t.currentCallFrame.CodeContractAbi != nil {
+		method, _ = t.currentCallFrame.CodeContractAbi.MethodById(t.currentCallFrame.InputData)
+	}
+
+	if method != nil {
+		// Unpack our output values and obtain a string to represent them, only if we didn't encounter an error.
+		if t.currentCallFrame.ReturnError == nil {
+			outputValue, _ := method.Outputs.Unpack(t.currentCallFrame.ReturnData)
+			outputValues = append(outputValues, outputValue)
+		}
+	}
+
+	return outputValues
+}
+
+// captureExitedCallFrame is a helper method used when a call frame is exited, to record information about it.
+func (v *ValueGenerationTracer) captureExitedCallFrame(output []byte, err error) {
+	// If this was an initial deployment, now that we're exiting, we'll want to record the finally deployed bytecodes.
+	if v.currentCallFrame.ToRuntimeBytecode == nil {
+		// As long as this isn't a failed contract creation, we should be able to fetch "to" byte code on exit.
+		if !v.currentCallFrame.IsContractCreation() || err == nil {
+			v.currentCallFrame.ToRuntimeBytecode = v.evm.StateDB.GetCode(v.currentCallFrame.ToAddress)
+		}
+	}
+	if v.currentCallFrame.CodeRuntimeBytecode == nil {
+		// Optimization: If the "to" and "code" addresses match, we can simply set our "code" already fetched "to"
+		// runtime bytecode.
+		if v.currentCallFrame.CodeAddress == v.currentCallFrame.ToAddress {
+			v.currentCallFrame.CodeRuntimeBytecode = v.currentCallFrame.ToRuntimeBytecode
+		} else {
+			v.currentCallFrame.CodeRuntimeBytecode = v.evm.StateDB.GetCode(v.currentCallFrame.CodeAddress)
+		}
+	}
+
+	// Resolve our contract definitions on the call frame data, if they have not been.
+	v.resolveCallFrameContractDefinitions(v.currentCallFrame)
+
+	// Set our information for this call frame
+	v.currentCallFrame.ReturnData = slices.Clone(output)
+	v.currentCallFrame.ReturnError = err
+
+	if v.currentCallFrame.ReturnError == nil {
+		// Grab return data of the call frame
+		v.trace.transactionOutputValues = append(v.trace.transactionOutputValues, v.getCallFrameReturnValue())
+	}
+
+	// We're exiting the current frame, so set our current call frame to the parent
+	v.currentCallFrame = v.currentCallFrame.ParentCallFrame
+}
+
+func (v *ValueGenerationTracer) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+	// TODO: look for RET opcode (for now try getting them from currentCallFrame.ReturnData)
+	// Execute all "on next capture state" events and clear them.
+	for _, eventHandler := range v.onNextCaptureState {
+		eventHandler()
+	}
+	v.onNextCaptureState = nil
+
```

Comment:
You will need to add the following code:
```suggestion
	// Now that we have executed some code, we have access to the VM scope. From this, we can populate more
	// information about our call frame. If this is a delegate or proxy call, the sender/to/code addresses should
	// be appropriately represented in this structure. The information populated earlier on frame enter represents
	// the raw call data, before delegate transformations are applied, etc.
	if !t.currentCallFrame.ExecutedCode {
		t.currentCallFrame.SenderAddress = scope.Contract.CallerAddress
		t.currentCallFrame.ToAddress = scope.Contract.Address()
		if scope.Contract.CodeAddr != nil {
			t.currentCallFrame.CodeAddress = *scope.Contract.CodeAddr
		}

		// Mark code as having executed in this scope, so we don't set these values again (as cheat codes may affect it).
		// We also want to know if a given call scope executed code, or simply represented a value transfer call.
		t.currentCallFrame.ExecutedCode = true
	}
```